dex/ws/wslink.go

```diff
@@ -124,57 +129,54 @@ func (c *WSLink) SendError(id uint64, rpcErr *msgjson.Error) {
 	}
 }
 
-// Start begins processing input and output messages.
-func (c *WSLink) Start() {
+// Connect begins processing input and output messages.
+func (c *WSLink) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	// Set the initial read deadline now that the ping ticker is about to be
 	// started. The pong handler will set subsequent read deadlines. 2x ping
 	// period is a very generous initial pong wait; the readWait provided to
 	// NewConnection could be stored and used here (once) instead.
+	if !atomic.CompareAndSwapUint32(&c.on, 0, 1) {
+		return nil, fmt.Errorf("Attempted to Start a running WSLink")
+	}
+	linkCtx, quit := context.WithCancel(ctx)
+	c.quit = quit
 	err := c.conn.SetReadDeadline(time.Now().Add(c.pingPeriod * 2))
 	if err != nil {
-		log.Errorf("Failed to set initial read deadline for %v: %v", c.ip, err)
-		return
+		return nil, fmt.Errorf("Failed to set initial read deadline for %v: %v", c.ip, err)
 	}
 
 	log.Tracef("Starting websocket messaging with peer %s", c.ip)
 	// Start processing input and output.
 	c.wg.Add(2)
-	go c.inHandler()
-	go c.outHandler()
+	go c.inHandler(linkCtx)
+	go c.pingHandler(linkCtx)
+	return &c.wg, nil
 }
 
 // Disconnect closes both the underlying websocket connection and the quit
 // channel.
 func (c *WSLink) Disconnect() {
-	c.quitMtx.Lock()
-	defer c.quitMtx.Unlock()
-	if !c.on {
+	if !atomic.CompareAndSwapUint32(&c.on, 1, 0) {
 		log.Debugf("Disconnect attempted on stopped WSLink.")
 		return
 	}
 	log.Tracef("Closing connection with peer %v", c.ip)
-	c.on = false
+	c.quit()
+	c.sendWG.Wait()
 	c.conn.Close()
-	close(c.quit)
-}
-
-// WaitForShutdown blocks until the WSLink goroutines are stopped and the
-// connection is closed.
-func (c *WSLink) WaitForShutdown() {
-	c.wg.Wait()
 }
 
 // inHandler handles all incoming messages for the websocket connection. It must
 // be run as a goroutine.
-func (c *WSLink) inHandler() {
+func (c *WSLink) inHandler(ctx context.Context) {
+	// Ensure the connection is closed.
+	defer c.Disconnect()
+	defer c.wg.Done()
```

Comment:
last in first out, so `defer c.Disconnect` should be after `defer c.wg.Done()` to ensure Disconnect gets called.